Documentation/design-docs/native-hosting-initialize.md

```diff
@@ -0,0 +1,213 @@
+# Native hosting with explicit initialize
+
+## Goals
+* All hosts should be able to use the new API (whether they will is a separate question as the old API has to be kept for backward compat reasons)
+* Hide implementation details as much as possible
+  * Make the API generally easier to use/understand
+  * Give the implementation more freedom
+  * Allow future improvements without breaking the API
+  * Consider explicitly documenting types of behaviors which nobody should take dependency on (specifically failure scenarios)
+* Extensible
+
+## New scenarios
+The API should allow these scenarios:
+* Runtime properties
+  * Specify additional runtime properties from the native host
+  * Implement conflict resolution for runtime properties
+  * Inspect calculated runtime properties (the ones calculated by `hostfxr`/`hostpolicy`)
+
+It should be possible to ship with only some of these supported, then enable more scenarios later on.
+
+## New `hostfxr` APIs
+
+Terminology:
+* "native host" - the code which uses the proposed APIs. Typically a native (C++) application, but can really be any "non .NET Core" code.
```

Comment:
>Typically a native (C++) application

In the spirit of precision this is a bad idea. Although we can use C++, all exposed function contracts must be marked `extern "C"` or specifically have C-style linking. Otherwise we are going to be exposed to compiler mangling sadness. Ideally we should define a "native host" more closely to the latter part of the sentence "can be any "non .NET Core" code." Alternatively or additionally we can specify that a "native host" is any application that explicitly activates CoreCLR through the hosting C-style API.

Reply:
Changed the wording to avoid specific languages and such.